{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled0.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 130
        },
        "id": "kYl1uYOitXCi",
        "outputId": "7bb3e36c-d88b-4450-a5c1-a723f7517ab2"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-0a81a9de02f2>\"\u001b[0;36m, line \u001b[0;32m188\u001b[0m\n\u001b[0;31m    print \"Q1: The alignments for naive match algorithm is %d\\n\"%patterns.naiveMatch(0)[1]\u001b[0m\n\u001b[0m                                                               ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ],
      "source": [
        "\n",
        "   \n",
        "# -*- coding: utf-8 -*-\n",
        "\"\"\"\n",
        "Created on Thu Oct 15 20:35:13 2015\n",
        "This project includes findPatternV2 class. The findPatternV2\n",
        "class finds the occurance and position of a given pattern in a given genomic \n",
        "sequence.\n",
        "@author: zhihuixie\n",
        "\"\"\"\n",
        "import bisect\n",
        "\n",
        "class findPatternV2 ():\n",
        "    \"\"\"\n",
        "    This class finds the occurance and position of a given pattern in a given \n",
        "    genomic sequence in a file.\n",
        "    \"\"\"\n",
        "    def __init__(self, pattern, filename = False):\n",
        "        # initiate parameters\n",
        "        self.pattern = pattern\n",
        "        self.filename = filename\n",
        "        \n",
        "    def readGenome (self):\n",
        "        \"\"\"\n",
        "        read genomic DNA sequence to a string\n",
        "        \"\"\"\n",
        "        genome = \"\"\n",
        "        with open (self.filename, \"r\") as f:\n",
        "            for line in f:\n",
        "                # skip header\n",
        "                if not line[0] == \">\":\n",
        "                    genome += line.rstrip()\n",
        "            f.close()\n",
        "        return genome\n",
        "        \n",
        "    def naiveMatch (self, numberOfMismatch, text = False):\n",
        "        \"\"\"\n",
        "        this is naive match to find the index of matched patterns in a genome\n",
        "        and calculate number of total character comparisons and alignments\n",
        "        \"\"\"\n",
        "        if text != False: #for test cases\n",
        "            genome = text\n",
        "        else:\n",
        "            genome = self.readGenome()\n",
        "        pattern = self.pattern\n",
        "        occurances = []\n",
        "        alignments = 0\n",
        "        comparisons = 0\n",
        "        for i in range(len(genome) - len(pattern) + 1):\n",
        "            match = True\n",
        "            counter = 0\n",
        "            for j in range(len(pattern)):\n",
        "                comparisons += 1\n",
        "                if pattern[j] != genome[i+j]:\n",
        "                    counter += 1\n",
        "                if counter > numberOfMismatch:\n",
        "                    match = False\n",
        "                    break\n",
        "            if match:\n",
        "                occurances.append(i)\n",
        "            alignments += 1\n",
        "        return occurances, alignments, comparisons\n",
        "        \n",
        "    def boyerMoore (self, numberOfMismatch, bm, text = False):\n",
        "        \"\"\"\n",
        "        this is naive match to find the index of matched patterns in a genome\n",
        "        and calculate number of total character comparisons and alignments\n",
        "        \"\"\"\n",
        "        i = 0\n",
        "        if text != False: #for test cases\n",
        "            genome = text\n",
        "        else:\n",
        "            genome = self.readGenome()\n",
        "        pattern = self.pattern\n",
        "        occurances = []\n",
        "        alignments = 0\n",
        "        comparisons = 0\n",
        "        while i < len(genome) - len(pattern) + 1:\n",
        "            shift = 1\n",
        "            match = True\n",
        "            for j in range(len(pattern) - 1, -1, -1):\n",
        "                comparisons += 1\n",
        "                if pattern[j] != genome[i+j]:\n",
        "                    badCharacterSkip = bm.bad_character_rule(j, genome[i+j])\n",
        "                    goodSuffixSkip = bm.good_suffix_rule(j)\n",
        "                    shift = max(shift, badCharacterSkip, goodSuffixSkip)\n",
        "                    match = False\n",
        "                    break\n",
        "            if match:\n",
        "                occurances.append(i)\n",
        "                goodSuffixSkip = bm.match_skip()\n",
        "                shift = max(shift, goodSuffixSkip)\n",
        "            i += shift\n",
        "            alignments += 1\n",
        "        return occurances, alignments, comparisons\n",
        "        \n",
        "    def matchedIndex(self, index, k_mer, pattern, isSubseqIndex):\n",
        "        \"\"\"\n",
        "        find number of hits, occurances and time of occurance for a given pattern\n",
        "        using string index in genome\n",
        "        \"\"\"\n",
        "        genome = self.readGenome()\n",
        "        occurances_match = []\n",
        "        hit_index = []\n",
        "        occurance_genome = []\n",
        "        counter = 0\n",
        "        if not isSubseqIndex:\n",
        "            length = len(pattern)-k_mer + 1\n",
        "        else:\n",
        "            length = isSubseqIndex\n",
        "        for i in range(length): # loop over to generate kmers\n",
        "            if not isSubseqIndex:                \n",
        "                pattern_q = pattern[i:i+k_mer]\n",
        "            else:\n",
        "                pattern_q = pattern[i:]\n",
        "            hits = index.query(pattern_q) # query each kmer\n",
        "            for hit in hits:\n",
        "                counter += 1 #count total number of hits\n",
        "                text = genome[hit-i : hit+len(pattern) -i]\n",
        "                if hit-i not in hit_index: #avoid duplicated counts\n",
        "                    hit_index.append(hit-i)\n",
        "                    occurance, _, _ = self.naiveMatch(2,text)\n",
        "                    occurances_match.extend(occurance)\n",
        "                if len(occurance) != 0 and hit-i not in occurance_genome:\n",
        "                    occurance_genome.append(hit-i) \n",
        "        return occurance_genome, len(occurances_match), counter\n",
        "        \n",
        "class Index(object):\n",
        "    def __init__(self, t, k):\n",
        "        ''' Create index from all substrings of size 'length' '''\n",
        "        self.k = k  # k-mer length (k)\n",
        "        self.index = []\n",
        "        for i in range(len(t) - k + 1):  # for each k-mer\n",
        "            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n",
        "        self.index.sort()  # alphabetize by k-mer\n",
        "    \n",
        "    def query(self, p):\n",
        "        ''' Return index hits for first k-mer of P '''\n",
        "        kmer = p[:self.k]  # query with first k-mer\n",
        "        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n",
        "        hits = []\n",
        "        while i < len(self.index):  # collect matching index entries\n",
        "            if self.index[i][0] != kmer:\n",
        "                break\n",
        "            hits.append(self.index[i][1])\n",
        "            i += 1\n",
        "        return hits\n",
        "    def genome_index(self):\n",
        "         return self.index       \n",
        "\n",
        "class SubseqIndex(object):\n",
        "    \"\"\" Holds a subsequence index for a text T \"\"\"\n",
        "    \n",
        "    def __init__(self, t, k, ival):\n",
        "        \"\"\" Create index from all subsequences consisting of k characters\n",
        "            spaced ival positions apart.  E.g., SubseqIndex(\"ATAT\", 2, 2)\n",
        "            extracts (\"AA\", 0) and (\"TT\", 1). \"\"\"\n",
        "        self.k = k  # num characters per subsequence extracted\n",
        "        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n",
        "        self.index = []\n",
        "        self.span = 1 + ival * (k - 1)\n",
        "        for i in range(len(t) - self.span + 1):  # for each subseq\n",
        "            self.index.append((t[i:i+self.span:ival], i))  # add (subseq, offset)\n",
        "        self.index.sort()  # alphabetize by subseq\n",
        "    \n",
        "    def query(self, p):\n",
        "        \"\"\" Return index hits for first subseq of p \"\"\"\n",
        "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
        "        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n",
        "        hits = []\n",
        "        while i < len(self.index):  # collect matching index entries\n",
        "            if self.index[i][0] != subseq:\n",
        "                break\n",
        "            hits.append(self.index[i][1])\n",
        "            i += 1\n",
        "        return hits\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    from bm_preproc import BoyerMoore\n",
        "    #Questions 1-3\n",
        "    filename = (\"../data/chr1.GRCh38.excerpt.fasta\")\n",
        "    #Q1: How many alignments does the naive exact matching algorithm try when \n",
        "    #matching the string GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
        "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
        "    #(Don't consider reverse complements.)\n",
        "    pattern = \"GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG\"\n",
        "    patterns = findPatternV2(pattern, filename)\n",
        "    print \"Q1: The alignments for naive match algorithm is %d\\n\"%patterns.naiveMatch(0)[1]\n",
        "    patterns = findPatternV2(\"GGCGCGGTGGCTCACGCCTGTAAT\", filename)    \n",
        "    \n",
        "    #Q2: How many character comparisons does the naive exact matching algorithm \n",
        "    #try when matching the string GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
        "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
        "    #(Don't consider reverse complements.)\n",
        "    print \"Q2: The characters comparisons for naive match algorithm is %d\\n\"%patterns.naiveMatch(0)[2]\n",
        "    \n",
        "    #How many alignments does Boyer-Moore try when matching the string \n",
        "    #GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG \n",
        "    #(derived from human Alu sequences) to the excerpt of human chromosome 1? \n",
        "    #(Don't consider reverse complements.)\n",
        "    print \"Q3: The alignments for Boyer-Moore algorithm is %d\\n\"%patterns.boyerMoore(0, \\\n",
        "          BoyerMoore(pattern, \"ACGT\"))[1]\n",
        "    \n",
        "    #Q4: How many times does the string GGCGCGGTGGCTCACGCCTGTAAT, \n",
        "    #which is derived from a human Alu sequence, occur with up to 2 \n",
        "    #substitutions in the excerpt of human chromosome 1? \n",
        "    #(Don't consider reverse complements here.)\n",
        "    k_mer = 8      \n",
        "    pattern = \"GGCGCGGTGGCTCACGCCTGTAAT\"\n",
        "    genome = patterns.readGenome()\n",
        "    index = Index(genome, k_mer)\n",
        "    occurances, numberOfOccurs, numberOfhits = patterns.matchedIndex(index, \\\n",
        "                                               k_mer, pattern, isSubseqIndex = False)\n",
        "    print \"Q4: Within 2 mismatchs, the string occurs %d times\\n\" %numberOfOccurs\n",
        "    \n",
        "    #Q5:Using the instructions given in Question 4, how many total index hits \n",
        "    #are there when searching for occurrences of GGCGCGGTGGCTCACGCCTGTAAT with \n",
        "    #up to 2 substitutions in the excerpt of human chromosome 1?\n",
        "    print \"Q5: Within 2 mismatchs, the total index hits are %d \\n\" %numberOfhits\n",
        "    \n",
        "    #Q6: When using this function, how many total index hits are there when \n",
        "    #searching for GGCGCGGTGGCTCACGCCTGTAAT with up to 2 substitutions in the \n",
        "    #excerpt of human chromosome 1? (Again, don't consider reverse complements.)\n",
        "    pattern = \"GGCGCGGTGGCTCACGCCTGTAAT\"\n",
        "    k_mer = 8\n",
        "    vial = 3\n",
        "    index = SubseqIndex(genome, k_mer, vial)\n",
        "    occurances, numberOfOccurs, numberOfhits = patterns.matchedIndex(index, \\\n",
        "                                               k_mer, pattern, isSubseqIndex = vial)\n",
        "\n",
        "    print \"Q6: Within 2 mismatchs, the hits are %d\\n\" %numberOfhits\n",
        "    "
      ]
    }
  ]
}